src/rust/engine/fs/store/src/lib.rs

```diff
@@ -951,55 +965,75 @@ impl Store {
     Ok(missing.is_empty())
   }
 
-  ///
-  /// Ensure that a directory is locally loadable, which will download it from the Remote store as
-  /// a sideeffect (if one is configured).
-  ///
-  pub async fn ensure_local_has_recursive_directory(
+  /// Ensure that the files are locally loadable. This will download them from the remote store as
+  /// a side effect, if one is configured.
+  pub async fn ensure_downloaded(
     &self,
-    dir_digest: DirectoryDigest,
+    mut file_digests: HashSet<Digest>,
+    directory_digests: HashSet<DirectoryDigest>,
   ) -> Result<(), StoreError> {
-    let mut file_digests = Vec::new();
-    self
-      .load_digest_trie(dir_digest)
-      .await?
-      .walk(SymlinkBehavior::Aware, &mut |_, entry| match entry {
-        directory::Entry::File(f) => file_digests.push(f.digest()),
-        directory::Entry::Symlink(_) | directory::Entry::Directory(_) => (),
-      });
+    let file_digests_from_directories =
+      future::try_join_all(directory_digests.into_iter().map(|dir_digest| async move {
+        let mut maybe_file_digest = None;
+        let trie = self.load_digest_trie(dir_digest).await?;
+        trie.walk(SymlinkBehavior::Aware, &mut |_, entry| match entry {
+          directory::Entry::File(f) => {
+            maybe_file_digest = Some(f.digest());
+          }
+          directory::Entry::Symlink(_) | directory::Entry::Directory(_) => (),
+        });
+        self.record_digest_trie(trie, true).await?;
```

Comment:
@stuhood this sets `initial_lease` to `true`. I wasn't sure what that was about

Reply:
A comment here to explain why we do this (or in the method signature) would be great.